File: _old/internal/extract/extract.go
// Package extract pulls values out of HTTP response bodies into named variables
// using a tiny rule format. Rules are written in the request's Scripts tab, one
// per line, in the form:
//
//	NAME = json.dotted.path
//	NAME = regex /pattern/        (first capture group)
//	NAME = header Some-Header     (response header value)
//
// Lines starting with # are comments. Empty lines are ignored. Unknown rule
// types are skipped.
package extract

import (
	"encoding/json"
	"net/http"
	"regexp"
	"strconv"
	"strings"
)

// Rule is one parsed extraction directive.
type Rule struct {
	Name string
	Kind string // "json", "regex", "header"
	Arg  string
}

// ParseRules parses the Scripts-tab text into a list of rules.
func ParseRules(script string) []Rule {
	var rules []Rule
	for _, line := range strings.Split(script, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		eq := strings.Index(line, "=")
		if eq < 0 {
			continue
		}
		name := strings.TrimSpace(line[:eq])
		rest := strings.TrimSpace(line[eq+1:])
		if name == "" || rest == "" {
			continue
		}
		fields := strings.SplitN(rest, " ", 2)
		if len(fields) != 2 {
			continue
		}
		kind := strings.ToLower(fields[0])
		arg := strings.TrimSpace(fields[1])
		switch kind {
		case "json", "regex", "header":
			rules = append(rules, Rule{Name: name, Kind: kind, Arg: arg})
		}
	}
	return rules
}

// Apply runs the rules against the given response body / headers and returns
// a map of variable name → extracted value. Failed extractions are omitted.
func Apply(rules []Rule, body string, headers http.Header) map[string]string {
	out := map[string]string{}
	var jsonRoot any
	jsonParsed := false

	for _, r := range rules {
		switch r.Kind {
		case "json":
			if !jsonParsed {
				_ = json.Unmarshal([]byte(body), &jsonRoot)
				jsonParsed = true
			}
			if v, ok := walkJSON(jsonRoot, r.Arg); ok {
				out[r.Name] = v
			}
		case "regex":
			pattern := strings.Trim(r.Arg, "/")
			re, err := regexp.Compile(pattern)
			if err != nil {
				continue
			}
			m := re.FindStringSubmatch(body)
			if len(m) >= 2 {
				out[r.Name] = m[1]
			} else if len(m) == 1 {
				out[r.Name] = m[0]
			}
		case "header":
			if headers != nil {
				if v := headers.Get(r.Arg); v != "" {
					out[r.Name] = v
				}
			}
		}
	}
	return out
}

// walkJSON resolves a dotted path like "data.items.0.id" against a parsed
// JSON value. Numeric segments index into arrays.
func walkJSON(root any, path string) (string, bool) {
	cur := root
	for _, seg := range strings.Split(path, ".") {
		if seg == "" {
			continue
		}
		switch v := cur.(type) {
		case map[string]any:
			next, ok := v[seg]
			if !ok {
				return "", false
			}
			cur = next
		case []any:
			idx, err := strconv.Atoi(seg)
			if err != nil || idx < 0 || idx >= len(v) {
				return "", false
			}
			cur = v[idx]
		default:
			return "", false
		}
	}
	return stringify(cur)
}

func stringify(v any) (string, bool) {
	switch x := v.(type) {
	case nil:
		return "", false
	case string:
		return x, true
	case bool:
		if x {
			return "true", true
		}
		return "false", true
	case float64:
		// JSON numbers come back as float64; render integers without trailing zero.
		if x == float64(int64(x)) {
			return strconv.FormatInt(int64(x), 10), true
		}
		return strconv.FormatFloat(x, 'g', -1, 64), true
	default:
		b, err := json.Marshal(v)
		if err != nil {
			return "", false
		}
		return string(b), true
	}
}

File: crates/binman-core/src/extract.rs
//! Pulling values out of a response into variables.
//!
//! Rules are written in a request's Scripts tab, one per line:
//!
//! ```text
//! token = json .access_token
//! id    = json data.items.0.id
//! csrf  = header X-CSRF-Token
//! code  = regex /code=(\d+)/
//! ```
//!
//! `#` starts a comment. A rule whose kind is not one of the three is passed
//! over rather than failing the others.

use regex::Regex;
use serde_json::Value;

use crate::request::header;
use crate::vars::Vars;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Rule {
    pub name: String,
    pub source: Source,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Source {
    /// A dotted path into the JSON body; numeric segments index arrays.
    Json(String),
    /// The first capture group of the pattern, or the whole match without one.
    Regex(String),
    Header(String),
}

pub fn parse(script: &str) -> Vec<Rule> {
    script
        .lines()
        .filter_map(|line| {
            let line = line.trim();
            if line.is_empty() || line.starts_with('#') {
                return None;
            }
            let (name, rest) = line.split_once('=')?;
            let (name, rest) = (name.trim(), rest.trim());
            if name.is_empty() || rest.is_empty() {
                return None;
            }
            let (kind, argument) = rest.split_once(' ')?;
            let argument = argument.trim().to_string();
            let source = match kind.to_ascii_lowercase().as_str() {
                "json" => Source::Json(argument),
                "regex" => Source::Regex(argument),
                "header" => Source::Header(argument),
                _ => return None,
            };
            Some(Rule {
                name: name.to_string(),
                source,
            })
        })
        .collect()
}

/// Runs the rules against a response. A rule that finds nothing is left out
/// rather than setting its variable to an empty string, which would shadow a
/// value from a lower layer.
pub fn apply(rules: &[Rule], body: &str, headers: &[(String, String)]) -> Vars {
    let mut json: Option<Option<Value>> = None;
    let mut out = Vars::new();

    for rule in rules {
        let found = match &rule.source {
            Source::Json(path) => {
                let root = json.get_or_insert_with(|| serde_json::from_str(body).ok());
                root.as_ref().and_then(|root| walk(root, path))
            }
            Source::Regex(pattern) => {
                Regex::new(pattern.trim_matches('/'))
                    .ok()
                    .and_then(|regex| {
                        let captures = regex.captures(body)?;
                        captures
                            .get(1)
                            .or_else(|| captures.get(0))
                            .map(|found| found.as_str().to_string())
                    })
            }
            Source::Header(name) => header(headers, name)
                .filter(|value| !value.is_empty())
                .map(str::to_string),
        };
        if let Some(value) = found {
            out.insert(rule.name.clone(), value);
        }
    }
    out
}

fn walk(root: &Value, path: &str) -> Option<String> {
    let mut current = root;
    for segment in path.split('.').filter(|segment| !segment.is_empty()) {
        current = match current {
            Value::Object(map) => map.get(segment)?,
            Value::Array(items) => items.get(segment.parse::<usize>().ok()?)?,
            _ => return None,
        };
    }
    match current {
        Value::Null => None,
        Value::String(text) => Some(text.clone()),
        Value::Bool(flag) => Some(flag.to_string()),
        Value::Number(number) => Some(number.to_string()),
        other => Some(other.to_string()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_the_three_kinds_and_skips_the_rest() {
        let rules = parse(
            "# header comment\n\
             TOKEN = json data.access_token\n\
             ID = regex /id=(\\d+)/\n\
             SERVER = header Server\n\
             malformed line\n\
             OTHER = xpath //a\n",
        );
        assert_eq!(rules.len(), 3, "{rules:?}");
        assert_eq!(
            rules[0],
            Rule {
                name: "TOKEN".into(),
                source: Source::Json("data.access_token".into())
            }
        );
    }

    #[test]
    fn walks_json_paths() {
        let rules = parse(
            "ID = json data.0.id\n\
             NAME = json data.0.name\n\
             ACTIVE = json data.0.active\n\
             TOKEN = json .token\n\
             MISSING = json data.99.id\n",
        );
        let out = apply(
            &rules,
            r#"{"token":"t","data":[{"id": 42, "name": "alice", "active": true}]}"#,
            &[],
        );
        assert_eq!(out["ID"], "42");
        assert_eq!(out["NAME"], "alice");
        assert_eq!(out["ACTIVE"], "true");
        assert_eq!(out["TOKEN"], "t");
        assert!(!out.contains_key("MISSING"));
    }

    #[test]
    fn a_regex_takes_its_first_group() {
        let out = apply(
            &parse("ID = regex /id=(\\d+)/"),
            "user id=4711 something",
            &[],
        );
        assert_eq!(out["ID"], "4711");
    }

    #[test]
    fn a_header_is_found_in_any_case() {
        let headers = vec![("x-trace-id".to_string(), "abc123".to_string())];
        let out = apply(&parse("TID = header X-Trace-Id"), "", &headers);
        assert_eq!(out["TID"], "abc123");
    }
}
